### PR TITLE
Reg struct copy.

### PR DESCRIPTION
**Pull request by sandreenko**
This PR changes the way we handle struct copies. 
https://github.com/dotnet/runtime/pull/32362/commits/cc8c36679c41b0929f03bc80da4cc16b2f568ef8: First, it deletes rationalize transformation that was converting `ASG(LCL_VAR struct, smth)` into `STORE_BLK` or `STORE_OBJ`.
https://github.com/dotnet/runtime/pull/32362/commits/3bc8f8e7c0dbe44450dee72134cfe78713b57c70: Then it skips `fgMorphOneAsgBlockOp` transformation for `ASG(LCL_VAR struct, smth)`, where `struct` can be replaced with a primitive type.

The overall diffs are positive, but small because of most of these structs are still marked as `doNotEnreg` in `impNormStructVal`, it will be fixed next.

Positive diffs come from:
1. using `movups` (for `STORE_LCL_VAR` instead of `movdqu` that is currently used for `STORE_BLK`), there was a PR https://github.com/dotnet/runtime/pull/1367/files that would use `movups` for all, but for now, it is useful for me to have different codegen for them, allows me to find diffs easily.
2. VN and CSE can propagate `ASG(struct, 0)` and similar constructions now, so we can coalesce chains of struct assignments now for zero-init, I will add non-zero cases later. It gives us diffs like `-29 (-53.70% of base) : System.Private.CoreLib.dasm - SyncSuccessSentinelStateMachineBox[VoidTaskResult][System.Threading.Tasks.VoidTaskResult]:.ctor():this`, that will be an issue in the next PR, see the discussion here https://github.com/dotnet/runtime/issues/1231#issuecomment-576477779;

The negative diffs come from:
1. `STORE_BLK` contained logic is less efficient, it often doesn't mark `Data` as contained when it is generating only one instruction;
2. `fgMorphOneAsgBlockOp` could retype structs with one gcField into primitive type copy without barriers and reporting, now we are generating costly `STORE_OBJ`, a primitive fix for it fails on a few library tests, I am checking them now.

<details>
 <summary>diffs for the Reg struct copies commit, x64, crossgen framework</summary>

```
Total bytes of diff: -4957 (-0.02% of base)
    diff is an improvement.
Top file improvements (bytes):
       -1274 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.04% of base)
       -1075 : System.Private.Xml.dasm (-0.03% of base)
        -496 : System.Linq.Expressions.dasm (-0.02% of base)
        -324 : System.Data.Common.dasm (-0.03% of base)
        -322 : System.Private.CoreLib.dasm (-0.01% of base)
        -308 : System.Transactions.Local.dasm (-0.30% of base)
        -179 : Newtonsoft.Json.dasm (-0.03% of base)
        -140 : Microsoft.CodeAnalysis.CSharp.dasm (-0.01% of base)
        -140 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.01% of base)
        -114 : System.Linq.dasm (-0.08% of base)
         -98 : System.Linq.Parallel.dasm (-0.02% of base)
         -76 : System.Security.AccessControl.dasm (-0.12% of base)
         -62 : xunit.performance.metrics.dasm (-0.46% of base)
         -56 : System.Private.DataContractSerialization.dasm (-0.01% of base)
         -54 : Microsoft.CodeAnalysis.dasm (-0.01% of base)
         -36 : System.IO.FileSystem.dasm (-0.06% of base)
         -35 : xunit.execution.dotnet.dasm (-0.02% of base)
         -28 : Microsoft.CSharp.dasm (-0.01% of base)
         -24 : System.Reflection.Metadata.dasm (-0.01% of base)
         -20 : System.Security.Cryptography.X509Certificates.dasm (-0.01% of base)
37 total files with Code Size differences (37 improved, 0 regressed), 72 unchanged.
Top method improvements (bytes):
        -304 (-0.91% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ClrTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
        -296 (-0.82% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ClrPrivateTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
        -256 (-0.53% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.KernelTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
         -74 (-0.07% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ApplicationServerTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
         -50 (-2.43% of base) : System.Private.Xml.dasm - System.Xml.Schema.Compiler:CalculateEffectiveTotalRange(System.Xml.Schema.XmlSchemaParticle,byref,byref):this
         -50 (-2.42% of base) : System.Private.Xml.dasm - System.Xml.Schema.SchemaCollectionCompiler:CalculateEffectiveTotalRange(System.Xml.Schema.XmlSchemaParticle,byref,byref):this
         -44 (-1.50% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.LocalRewriter:MakeDecimalLiteral(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode,Microsoft.CodeAnalysis.ConstantValue):Microsoft.CodeAnalysis.CSharp.BoundExpression:this
         -44 (-2.42% of base) : System.Data.Common.dasm - System.Data.Common.TimeSpanStorage:Aggregate(System.Int32[],int):System.Object:this
         -40 (-0.87% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionEvaluator:PerformCompileTimeBinaryOperation(ushort,byte,Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst,Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst,Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax):Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst
         -40 (-1.66% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ConvertUtils:DecimalTryParse(System.Char[],int,int,byref):int
         -32 (-3.02% of base) : System.Private.Xml.dasm - System.Xml.Schema.SchemaCollectionCompiler:CalculateSequenceRange(System.Xml.Schema.XmlSchemaSequence,byref,byref):this
         -30 (-2.48% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.StartStopActivityComputer:FixAndProcessWindowsASP(Microsoft.Diagnostics.Tracing.TraceEvent,System.Collections.Generic.KeyValuePair`2[System.Guid,System.Guid][]):this
         -30 (-3.35% of base) : System.Private.Xml.dasm - System.Xml.Schema.Compiler:IsSequenceFromChoice(System.Xml.Schema.XmlSchemaSequence,System.Xml.Schema.XmlSchemaChoice):bool:this
         -30 (-1.68% of base) : System.Private.Xml.dasm - System.Xml.Xsl.IlGen.XmlILOptimizerVisitor:FoldArithmetic(int,System.Xml.Xsl.Qil.QilLiteral,System.Xml.Xsl.Qil.QilLiteral):System.Xml.Xsl.Qil.QilNode:this
         -26 (-0.66% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Binder:DoUncheckedConversion(byte,Microsoft.CodeAnalysis.ConstantValue):System.Object
         -26 (-0.57% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.Scanner:ScanNumericLiteral(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SyntaxList`1[[Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode, Microsoft.CodeAnalysis.VisualBasic, Version=1.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]):Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SyntaxToken:this
         -26 (-2.50% of base) : System.Private.CoreLib.dasm - System.TimeZoneInfo:CreateAdjustmentRuleFromTimeZoneInformation(byref,System.DateTime,System.DateTime,int):AdjustmentRule
         -24 (-0.37% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Binder:FoldNeverOverflowBinaryOperators(int,Microsoft.CodeAnalysis.ConstantValue,Microsoft.CodeAnalysis.ConstantValue):System.Object
         -24 (-2.11% of base) : Microsoft.CSharp.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.ExpressionBinder:BindDecimalConstCast(Microsoft.CSharp.RuntimeBinder.Semantics.CType,Microsoft.CSharp.RuntimeBinder.Semantics.CType,Microsoft.CSharp.RuntimeBinder.Semantics.ExprConstant):Microsoft.CSharp.RuntimeBinder.Semantics.Expr
         -24 (-1.17% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - <>c__DisplayClass0_0:<.ctor>b__1(Microsoft.Diagnostics.Tracing.TraceEvent):this
```
</details>

```
Total diffs for frameworks:
x64 Total bytes of diff: -5890 (-0.02% of base)
1338 total methods with Code Size differences (1262 improved, 76 regressed), 178158 unchanged.

arm64 Total bytes of diff: -23656 (-0.02% of base)
489 total methods with Code Size differences (311 improved, 178 regressed), 157900 unchanged.

arm Total bytes of diff: -1576 (-0.00% of base)
250 total methods with Code Size differences (169 improved, 81 regressed), 158053 unchanged.

x86 Total bytes of diff: -2962 (-0.01% of base)
580 total methods with Code Size differences (535 improved, 45 regressed), 178824 unchanged.
```

Based on https://github.com/dotnet/runtime/issues/1231#issuecomment-570156628.

Side note: `varTypeIsSmall` and similar functions in `vartype.h` should assert on `TYP_STRUCT` because they don't have enough information to get a correct result, the callers should be responsible for handling that. It is a preexisting issue, but that change bumps its priority, I am working on a fix in another branch.

### CODE REVIEW

**Comment by sandreenko** [comment]
There are 3 types of regressions with that PR right now:
1. `STORE_BLK` is not optimized for 1 instruction UNROLL case, so it does not set contained on IMM in cases like:
```
N001 (  1,  1) [000002] ------------         t2 =    CNS_INT   int    0 $41
               [000008] Dc-----N----         t8 =    LCL_VAR_ADDR byref  V01 loc0         
                                                  /--*  t8     byref  
                                                  +--*  t2     int    
N003 (  5,  4) [000003] sA----------              *  STORE_BLK struct<YieldAwaiter, 1> (init) (Unroll)
```
so we end up generating extra instruction;
2. bad local variable assignments, `Final local variable assignments` choose worse order than it was choosing before, so it has to generate bigger offsets after.
3. additional shadow variables, it is described here:
https://github.com/dotnet/runtime/blob/9d9a55a844330922b00bdbde211807eba5ca8455/src/coreclr/src/jit/gschecks.cpp#L239-L240

**Comment by sandreenko** [comment]
PTAL @CarolEidt @dotnet/jit-contrib

File: src/coreclr/src/jit/lower.cpp [hunk 1]
```diff
@@ -299,35 +299,58 @@ GenTree* Lowering::LowerNode(GenTree* node)
...
+                }
+                else
+#endif // FEATURE_MULTIREG_RET
+                    if (!src->OperIs(GT_LCL_VAR) || varDsc->GetLayout()->GetRegisterType() == TYP_UNDEF)
+                {
+                    GenTreeLclVar* addr = comp->gtNewLclVarAddrNode(store->GetLclNum(), TYP_BYREF);
+
+                    addr->gtFlags |= GTF_VAR_DEF;
+                    assert(!addr->IsPartialLclFld(comp));
+                    addr->gtFlags |= GTF_DONT_CSE;
```
Comment (sandreenko) [inline src/coreclr/src/jit/lower.cpp]: That is loweing, so nobody will read these flags, but I set them to have closer diffs.

**Review by CarolEidt — APPROVED** [review]
LGTM

File: src/coreclr/src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -4638,15 +4638,24 @@ void CodeGen::genCodeForStoreLclVar(GenTreeLclVar* tree)
...
         unsigned   lclNum = tree->GetLclNum();
-        LclVarDsc* varDsc = &(compiler->lvaTable[lclNum]);
+        LclVarDsc* varDsc = compiler->lvaGetDesc(lclNum);
+
+        var_types targetType = varDsc->GetRegisterType(tree);
 
-        // Ensure that lclVar nodes are typed correctly.
-        assert(!varDsc->lvNormalizeOnStore() || (targetType == genActualType(varDsc->TypeGet())));
+#ifdef DEBUG
+        var_types op1Type = op1->TypeGet();
```
Comment (sandreenko) [inline src/coreclr/src/jit/codegenxarch.cpp]: Merged my changes with the new assert from #32750.

File: src/coreclr/src/jit/morph.cpp [hunk 3]
```diff
@@ -8652,58 +8652,68 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
...
+        //          /   \.
+        //      GT_IND GT_IND or CNS_INT
+        //         |      |
+        //       [dest] [src]
+        //
+
+        if (asgType == TYP_STRUCT)
         {
-            if (clsHnd == NO_CLASS_HANDLE)
+            // It is possible to use `initobj` to init a primitive type on the stack,
```
Comment (sandreenko) [inline src/coreclr/src/jit/morph.cpp]: @kunalspathak thanks for your feedback, I added an example where `var` is not a struct, but `ASG` is.

**Pull request merged**